<a href="https://colab.research.google.com/github/Raudhil/Machine-Learning-2025/blob/main/Jobsheet%2013/RAUDHIL_FIRDAUS_NAUFAL_JS13_Praktikum3.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **Praktikum 1**

**Langkah:**

Buat dataset sederhana (XOR).

Inisialisasi bobot dan bias.

Implementasikan forward pass.

Hitung error dan lakukan backpropagation.

Update bobot menggunakan gradient descent.

In [ ]:
import numpy as np

# Dataset XOR
X = np.array([[0,0],[0,1],[1,0],[1,1]])
y = np.array([[0],[1],[1],[0]])

# Parameter
input_size = 2
hidden_size = 2
output_size = 1
lr = 0.1

# Inisialisasi bobot
W1 = np.random.randn(input_size, hidden_size)
b1 = np.zeros((1, hidden_size))
W2 = np.random.randn(hidden_size, output_size)
b2 = np.zeros((1, output_size))

# Fungsi aktivasi
def sigmoid(x):
    return 1 / (1 + np.exp(-x))

def sigmoid_derivative(x):
    return x * (1 - x)

# Training
for epoch in range(10000):
    # Forward pass
    z1 = np.dot(X, W1) + b1
    a1 = sigmoid(z1)
    z2 = np.dot(a1, W2) + b2
    a2 = sigmoid(z2)

    # Hitung error
    error = y - a2

    # Backpropagation
    d_a2 = error * sigmoid_derivative(a2)
    d_W2 = np.dot(a1.T, d_a2)
    d_b2 = np.sum(d_a2, axis=0, keepdims=True)

    d_a1 = np.dot(d_a2, W2.T) * sigmoid_derivative(a1)
    d_W1 = np.dot(X.T, d_a1)
    d_b1 = np.sum(d_a1, axis=0, keepdims=True)

    # Update bobot
    W1 += lr * d_W1
    b1 += lr * d_b1
    W2 += lr * d_W2
    b2 += lr * d_b2

    if epoch % 1000 == 0:
        loss = np.mean(np.square(error))
        print(f"Epoch {epoch}, Loss: {loss}")

# Output akhir
print("Prediksi:")
print(a2)

Epoch 0, Loss: 0.2540418689873392
Epoch 1000, Loss: 0.2489254088154776
Epoch 2000, Loss: 0.23535178772453808
Epoch 3000, Loss: 0.18195854197618286
Epoch 4000, Loss: 0.15147045564123343
Epoch 5000, Loss: 0.13993153803281277
Epoch 6000, Loss: 0.1348854364974286
Epoch 7000, Loss: 0.13221768204053658
Epoch 8000, Loss: 0.13060702516584713
Epoch 9000, Loss: 0.12954345910055642
Prediksi:
[[0.04847747]
 [0.50001619]
 [0.93245302]
 [0.50822002]]


**Tugas 1**


Ubah jumlah neuron hidden layer menjadi 3.

In [ ]:
# Dataset XOR
X = np.array([[0,0],[0,1],[1,0],[1,1]])
y = np.array([[0],[1],[1],[0]])

# Parameter
input_size = 2
hidden_size = 3
output_size = 1
lr = 0.1

# Inisialisasi bobot
W1 = np.random.randn(input_size, hidden_size)
b1 = np.zeros((1, hidden_size))
W2 = np.random.randn(hidden_size, output_size)
b2 = np.zeros((1, output_size))

# Fungsi aktivasi
def sigmoid(x):
    return 1 / (1 + np.exp(-x))

def sigmoid_derivative(x):
    return x * (1 - x)

# Training
for epoch in range(10000):
    # Forward pass
    z1 = np.dot(X, W1) + b1
    a1 = sigmoid(z1)
    z2 = np.dot(a1, W2) + b2
    a2 = sigmoid(z2)

    # Hitung error
    error = y - a2

    # Backpropagation
    d_a2 = error * sigmoid_derivative(a2)
    d_W2 = np.dot(a1.T, d_a2)
    d_b2 = np.sum(d_a2, axis=0, keepdims=True)

    d_a1 = np.dot(d_a2, W2.T) * sigmoid_derivative(a1)
    d_W1 = np.dot(X.T, d_a1)
    d_b1 = np.sum(d_a1, axis=0, keepdims=True)

    # Update bobot
    W1 += lr * d_W1
    b1 += lr * d_b1
    W2 += lr * d_W2
    b2 += lr * d_b2

    if epoch % 1000 == 0:
        loss = np.mean(np.square(error))
        print(f"Epoch {epoch}, Loss: {loss}")

# Output akhir
print("Prediksi:")
print(a2)

Epoch 0, Loss: 0.2822343450251595
Epoch 1000, Loss: 0.24998175281374235
Epoch 2000, Loss: 0.24846388663387
Epoch 3000, Loss: 0.2325788958924659
Epoch 4000, Loss: 0.10203055464812211
Epoch 5000, Loss: 0.021941342678303277
Epoch 6000, Loss: 0.009735466937629689
Epoch 7000, Loss: 0.0059380428412692645
Epoch 8000, Loss: 0.004190087796248304
Epoch 9000, Loss: 0.0032069833036875483
Prediksi:
[[0.03020431]
 [0.94855612]
 [0.94878151]
 [0.06445946]]


Bandingkan hasil loss dengan konfigurasi awal.



Berdasarkan hasil loss antara konfigurasi awal dengan konfigurasi baru yang menambahkan hidden neuron sebanyak 1, terlihat kalau loss turun hingga 0.0025, walaupun itu kecil, tapi bisa diketahui kalau menambah neuron dapat membantu menurunkan error.

Tambahkan fungsi aktivasi ReLU dan bandingkan hasil.



In [ ]:
import numpy as np

# Dataset XOR
X = np.array([[0,0],[0,1],[1,0],[1,1]])
y = np.array([[0],[1],[1],[0]])

# Hyperparams
input_size = 2
hidden_size = 3
output_size = 1
lr = 0.1
epochs = 10000
np.random.seed(42)

# initialize once and copy to ensure fair comparison
W1_init = np.random.randn(input_size, hidden_size)
b1_init = np.zeros((1, hidden_size))
W2_init = np.random.randn(hidden_size, output_size)
b2_init = np.zeros((1, output_size))

# Aktivasi
def sigmoid(x):
    return 1 / (1 + np.exp(-x))

def sigmoid_derivative(a):
    return a * (1 - a)

def relu(x):
    return np.maximum(0, x)

def relu_derivative(x):
    return (x > 0).astype(float)

# Train with sigmoid hidden
W1 = W1_init.copy(); b1 = b1_init.copy(); W2 = W2_init.copy(); b2 = b2_init.copy()
for epoch in range(epochs):
    z1 = X.dot(W1) + b1
    a1 = sigmoid(z1)
    z2 = a1.dot(W2) + b2
    a2 = sigmoid(z2)

    error = y - a2
    d_a2 = error * sigmoid_derivative(a2)
    d_W2 = a1.T.dot(d_a2)
    d_b2 = np.sum(d_a2, axis=0, keepdims=True)

    d_a1 = d_a2.dot(W2.T) * sigmoid_derivative(a1)
    d_W1 = X.T.dot(d_a1)
    d_b1 = np.sum(d_a1, axis=0, keepdims=True)

    W1 += lr * d_W1; b1 += lr * d_b1
    W2 += lr * d_W2; b2 += lr * d_b2

sigmoid_loss = np.mean(np.square(y - a2))
sigmoid_pred = a2.copy()

# Train with ReLU hidden (same init)
W1 = W1_init.copy(); b1 = b1_init.copy(); W2 = W2_init.copy(); b2 = b2_init.copy()
for epoch in range(epochs):
    z1 = X.dot(W1) + b1
    a1 = relu(z1)
    z2 = a1.dot(W2) + b2
    a2 = sigmoid(z2)

    error = y - a2
    d_a2 = error * sigmoid_derivative(a2)
    d_W2 = a1.T.dot(d_a2)
    d_b2 = np.sum(d_a2, axis=0, keepdims=True)

    d_a1 = d_a2.dot(W2.T) * relu_derivative(z1)
    d_W1 = X.T.dot(d_a1)
    d_b1 = np.sum(d_a1, axis=0, keepdims=True)

    W1 += lr * d_W1; b1 += lr * d_b1
    W2 += lr * d_W2; b2 += lr * d_b2

relu_loss = np.mean(np.square(y - a2))
relu_pred = a2.copy()

print("Sigmoid hidden final loss:", sigmoid_loss)
print("Sigmoid hidden predictions:\n", np.round(sigmoid_pred, 3))
print("ReLU hidden final loss:", relu_loss)
print("ReLU hidden predictions:\n", np.round(relu_pred, 3))
# ...existing code...

Sigmoid hidden final loss: 0.0022974157268842183
Sigmoid hidden predictions:
 [[0.025]
 [0.953]
 [0.951]
 [0.063]]
ReLU hidden final loss: 0.00035154169220683637
ReLU hidden predictions:
 [[0.03 ]
 [0.986]
 [0.986]
 [0.012]]


# **Praktikum 2**

**Langkah**

Import library.

Load dataset.

Bangun model.

Kompilasi dan latih model.

Evaluasi hasil.

In [ ]:
from sklearn.datasets import load_iris
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import OneHotEncoder
import tensorflow as tf
import numpy as np

# Load dataset
iris = load_iris()
X = iris.data
y = iris.target.reshape(-1, 1)

# One-hot encoding
encoder = OneHotEncoder(sparse_output=False) # Changed sparse=False to sparse_output=False
y = encoder.fit_transform(y)

# Split data
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2)

# Bangun model
model = tf.keras.Sequential([
    tf.keras.layers.Dense(10, activation='relu', input_shape=(4,)),
    tf.keras.layers.Dense(8, activation='relu'),
    tf.keras.layers.Dense(3, activation='softmax')
])

# Kompilasi
model.compile(optimizer='adam', loss='categorical_crossentropy', metrics=['accuracy'])

# Latih model
model.fit(X_train, y_train, epochs=50, batch_size=8)

# Evaluasi
loss, acc = model.evaluate(X_test, y_test)
print(f"Akurasi: {acc}")

Epoch 1/50


/usr/local/lib/python3.12/dist-packages/keras/src/layers/core/dense.py:93: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


15/15 ━━━━━━━━━━━━━━━━━━━━ 2s 12ms/step - accuracy: 0.3581 - loss: 1.7447
Epoch 2/50
15/15 ━━━━━━━━━━━━━━━━━━━━ 0s 12ms/step - accuracy: 0.3551 - loss: 1.2693
Epoch 3/50
15/15 ━━━━━━━━━━━━━━━━━━━━ 0s 12ms/step - accuracy: 0.5083 - loss: 0.9964
Epoch 4/50
15/15 ━━━━━━━━━━━━━━━━━━━━ 0s 16ms/step - accuracy: 0.6720 - loss: 0.9124
Epoch 5/50
15/15 ━━━━━━━━━━━━━━━━━━━━ 0s 10ms/step - accuracy: 0.6390 - loss: 0.8943
Epoch 6/50
15/15 ━━━━━━━━━━━━━━━━━━━━ 0s 19ms/step - accuracy: 0.6387 - loss: 0.8566
Epoch 7/50
15/15 ━━━━━━━━━━━━━━━━━━━━ 1s 21ms/step - accuracy: 0.6671 - loss: 0.7902
Epoch 8/50
15/15 ━━━━━━━━━━━━━━━━━━━━ 0s 5ms/step - accuracy: 0.5866 - loss: 0.8356  
Epoch 9/50
15/15 ━━━━━━━━━━━━━━━━━━━━ 0s 5ms/step - accuracy: 0.6009 - loss: 0.7947 
Epoch 10/50
15/15 ━━━━━━━━━━━━━━━━━━━━ 0s 5ms/step - accuracy: 0.7059 - loss: 0.6857 
Epoch 11/50
15/15 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - accuracy: 0.7144 - loss: 0.6476 
Epoch 12/50
15/15 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - accuracy: 0.7049 - l

**Tugas 2**



Ubah jumlah neuron hidden layer.


In [ ]:
# Load dataset
iris = load_iris()
X = iris.data
y = iris.target.reshape(-1, 1)

# One-hot encoding
encoder = OneHotEncoder(sparse_output=False) # Changed sparse=False to sparse_output=False
y = encoder.fit_transform(y)

# Split data
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2)

# Bangun model
model = tf.keras.Sequential([
    tf.keras.layers.Dense(16, activation='relu', input_shape=(4,)),
    tf.keras.layers.Dense(12, activation='relu'),
    tf.keras.layers.Dense(3, activation='softmax')
])

# Kompilasi
model.compile(optimizer='adam', loss='categorical_crossentropy', metrics=['accuracy'])

# Latih model
model.fit(X_train, y_train, epochs=50, batch_size=8)

# Evaluasi
loss, acc = model.evaluate(X_test, y_test)
print(f"Akurasi: {acc}")

Epoch 1/50


/usr/local/lib/python3.12/dist-packages/keras/src/layers/core/dense.py:93: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


15/15 ━━━━━━━━━━━━━━━━━━━━ 2s 3ms/step - accuracy: 0.0163 - loss: 1.8611    
Epoch 2/50
15/15 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - accuracy: 0.0121 - loss: 1.6412     
Epoch 3/50
15/15 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - accuracy: 0.0000e+00 - loss: 1.4430 
Epoch 4/50
15/15 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - accuracy: 0.0000e+00 - loss: 1.2985 
Epoch 5/50
15/15 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - accuracy: 0.0309 - loss: 1.1975     
Epoch 6/50
15/15 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - accuracy: 0.1373 - loss: 1.1000     
Epoch 7/50
15/15 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - accuracy: 0.6340 - loss: 1.0113 
Epoch 8/50
15/15 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - accuracy: 0.8487 - loss: 0.9301 
Epoch 9/50
15/15 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - accuracy: 0.9211 - loss: 0.8542 
Epoch 10/50
15/15 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - accuracy: 0.8566 - loss: 0.7825 
Epoch 11/50
15/15 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - accuracy: 0.8945 - loss: 0.7214 
Epoch 12/50
15/15 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step 

Bandingkan akurasi dengan konfigurasi awal.

Dari perbandingan ini, kita dapat melihat dengan menambah jumlah neuron pada hidden layer mendapatkan peningkatan akurasi pada model untuk dataset Iris.

**Tugas 3**


Bandingkan Sigmoid vs ReLU pada dataset Iris.



In [ ]:
from sklearn.datasets import load_iris
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import OneHotEncoder
import tensorflow as tf
import numpy as np

iris = load_iris()
X = iris.data
y = iris.target.reshape(-1, 1)

encoder = OneHotEncoder(sparse_output=False)
y = encoder.fit_transform(y)

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

def build_model(activation_func):
    model = tf.keras.Sequential([
        tf.keras.layers.Dense(10, activation=activation_func, input_shape=(4,)),
        tf.keras.layers.Dense(8, activation=activation_func),
        tf.keras.layers.Dense(3, activation='softmax') # Output layer always softmax for multi-class
    ])
    model.compile(optimizer='adam', loss='categorical_crossentropy', metrics=['accuracy'])
    return model

print("\nTraining model with Sigmoid activation...")
sigmoid_model = build_model('sigmoid')
sigmoid_history = sigmoid_model.fit(X_train, y_train, epochs=50, batch_size=8, verbose=0)
sigmoid_loss, sigmoid_acc = sigmoid_model.evaluate(X_test, y_test, verbose=0)

print("\nTraining model with ReLU activation...")
relu_model = build_model('relu')
relu_history = relu_model.fit(X_train, y_train, epochs=50, batch_size=8, verbose=0)
relu_loss, relu_acc = relu_model.evaluate(X_test, y_test, verbose=0)

print("\n--- Comparison of Sigmoid vs ReLU on Iris Dataset ---")
print(f"Sigmoid Model - Loss: {sigmoid_loss:.4f}, Accuracy: {sigmoid_acc:.4f}")
print(f"ReLU Model    - Loss: {relu_loss:.4f}, Accuracy: {relu_acc:.4f}")


Training model with Sigmoid activation...


/usr/local/lib/python3.12/dist-packages/keras/src/layers/core/dense.py:93: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)



Training model with ReLU activation...

--- Comparison of Sigmoid vs ReLU on Iris Dataset ---
Sigmoid Model - Loss: 0.6255, Accuracy: 0.8000
ReLU Model    - Loss: 0.2105, Accuracy: 0.9333


Catat perbedaan loss dan akurasi.



# **Praktikum 3**

In [ ]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
import tensorflow as tf

# Contoh dataset (buat dummy data)
data = pd.DataFrame({
    'luas': [50, 60, 70, 80, 90],
    'harga': [500, 600, 700, 800, 900]
})

X = data[['luas']]
y = data[['harga']]

# Normalisasi
scaler = StandardScaler()
X = scaler.fit_transform(X)
y = scaler.fit_transform(y)

# Split data
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2)

# Model
model = tf.keras.Sequential([
    tf.keras.layers.Dense(10, activation='relu', input_shape=(1,)),
    tf.keras.layers.Dense(1)
])

model.compile(optimizer='adam', loss='mse')
model.fit(X_train, y_train, epochs=100)

# Evaluasi
print("Prediksi:", model.predict(X_test))

Epoch 1/100


/usr/local/lib/python3.12/dist-packages/keras/src/layers/core/dense.py:93: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


1/1 ━━━━━━━━━━━━━━━━━━━━ 1s 1s/step - loss: 0.7557
Epoch 2/100
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 92ms/step - loss: 0.7436
Epoch 3/100
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 180ms/step - loss: 0.7315
Epoch 4/100
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 246ms/step - loss: 0.7196
Epoch 5/100
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 48ms/step - loss: 0.7077
Epoch 6/100
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 46ms/step - loss: 0.6960
Epoch 7/100
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 43ms/step - loss: 0.6844
Epoch 8/100
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 45ms/step - loss: 0.6729
Epoch 9/100
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 48ms/step - loss: 0.6616
Epoch 10/100
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 46ms/step - loss: 0.6504
Epoch 11/100
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 62ms/step - loss: 0.6393
Epoch 12/100
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 46ms/step - loss: 0.6283
Epoch 13/100
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 46ms/step - loss: 0.6175
Epoch 14/100
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 48ms/step - loss: 0.6068
Epoch 15/100
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 46ms/step - loss: 0.5962
Epoch 16/100
1/1 ━━━━━━━━━━━━━━

**Tugas 4:**



Ubah learning rate.



In [ ]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
import tensorflow as tf

# Contoh dataset (buat dummy data)
data = pd.DataFrame({
    'luas': [50, 60, 70, 80, 90],
    'harga': [500, 600, 700, 800, 900]
})

X = data[['luas']]
y = data[['harga']]

# Normalisasi
scaler = StandardScaler()
X = scaler.fit_transform(X)
y = scaler.fit_transform(y)

# Split data
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2)

# Model
model = tf.keras.Sequential([
    tf.keras.layers.Dense(10, activation='relu', input_shape=(1,)),
    tf.keras.layers.Dense(1)
])

# Ubah learning rate
custom_learning_rate = 0.01
optimizer = tf.keras.optimizers.Adam(learning_rate=custom_learning_rate)

model.compile(optimizer=optimizer, loss='mse')
model.fit(X_train, y_train, epochs=100)

# Evaluasi
print("Prediksi:", model.predict(X_test))

Epoch 1/100


/usr/local/lib/python3.12/dist-packages/keras/src/layers/core/dense.py:93: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


1/1 ━━━━━━━━━━━━━━━━━━━━ 1s 711ms/step - loss: 1.5003
Epoch 2/100
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 44ms/step - loss: 1.3449
Epoch 3/100
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 48ms/step - loss: 1.2029
Epoch 4/100
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 46ms/step - loss: 1.0780
Epoch 5/100
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 47ms/step - loss: 0.9638
Epoch 6/100
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 49ms/step - loss: 0.8597
Epoch 7/100
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 46ms/step - loss: 0.7668
Epoch 8/100
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 60ms/step - loss: 0.6819
Epoch 9/100
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 48ms/step - loss: 0.6035
Epoch 10/100
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 50ms/step - loss: 0.5314
Epoch 11/100
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 45ms/step - loss: 0.4682
Epoch 12/100
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 47ms/step - loss: 0.4099
Epoch 13/100
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 44ms/step - loss: 0.3565
Epoch 14/100
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 45ms/step - loss: 0.3077
Epoch 15/100
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 44ms/step - loss: 0.2634
Epoch 16/100
1/1 ━━━━━━━━━━━━━

Bandingkan hasil loss.

Perbandingan hasil loss:
Asli: loss sekitar 0.1324.

Modifikasi:loss sekitar 0.0054.

Dari hasil ini, jelas bahwa perubahan learning rate menjadi 0.01 menghasilkan loss akhir yang jauh lebih rendah. Hal ini mengindikasikan bahwa model berkonvergensi lebih baik dan mencapai prediksi yang jauh lebih akurat pada data pelatihan dengan learning rate yang baru ini.

# **Tugas Praktikum**


Load dataset MNIST dari Keras.

Bangun model dengan 2 hidden layer.

Latih model dan evaluasi akurasi.

In [ ]:
import tensorflow as tf
from tensorflow.keras.datasets import mnist
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense, Flatten
from tensorflow.keras.utils import to_categorical
import time

# 1. Load dataset MNIST
(X_train, y_train), (X_test, y_test) = mnist.load_data()

# Normalisasi data (0-255 → 0-1)
X_train = X_train / 255.0
X_test = X_test / 255.0

# One-hot encoding label
y_train = to_categorical(y_train, 10)
y_test = to_categorical(y_test, 10)

# 2. Bangun model JST
model = Sequential([
    Flatten(input_shape=(28, 28)), # Ubah gambar 28x28 menjadi vektor
    Dense(256, activation='relu'), # Hidden layer 1 dengan 256 neuron
    Dense(128, activation='relu'), # Hidden layer 2 dengan 128 neuron
    Dense(10, activation='softmax') # Output layer (10 kelas)
])

# 3. Kompilasi model
model.compile(optimizer='adam', loss='categorical_crossentropy', metrics=['accuracy'])

# 4. Latih model
print("\nTraining Model...")
start_time = time.time()
model.fit(X_train, y_train, epochs=10, batch_size=32, verbose=1)
end_time = time.time()
training_time = end_time - start_time
print(f"Training selesai dalam {training_time:.2f} detik")

# 5. Evaluasi model
print("\nEvaluating Model...")
loss, acc = model.evaluate(X_test, y_test, verbose=0)
print(f"Akurasi pada data uji: {acc:.4f}")


/usr/local/lib/python3.12/dist-packages/keras/src/layers/reshaping/flatten.py:37: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(**kwargs)



Training Model...
Epoch 1/10
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 21s 10ms/step - accuracy: 0.8904 - loss: 0.3683
Epoch 2/10
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 14s 8ms/step - accuracy: 0.9733 - loss: 0.0886
Epoch 3/10
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 15s 8ms/step - accuracy: 0.9814 - loss: 0.0576
Epoch 4/10
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 14s 8ms/step - accuracy: 0.9865 - loss: 0.0416
Epoch 5/10
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 14s 8ms/step - accuracy: 0.9899 - loss: 0.0309
Epoch 6/10
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 15s 8ms/step - accuracy: 0.9920 - loss: 0.0239
Epoch 7/10
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 16s 8ms/step - accuracy: 0.9935 - loss: 0.0204
Epoch 8/10
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 16s 9ms/step - accuracy: 0.9938 - loss: 0.0184
Epoch 9/10
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 15s 8ms/step - accuracy: 0.9943 - loss: 0.0174
Epoch 10/10
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 15s 8ms/step - accuracy: 0.9941 - loss: 0.0171
Training selesai dalam 161.50 detik

Evaluating Model...
Akurasi pada data uji: 0.9787



Ubah jumlah neuron di hidden layer (misal: 256 dan 128).

Tambahkan satu hidden layer lagi.

Bandingkan akurasi dan waktu pelatihan.

In [3]:
import tensorflow as tf
from tensorflow.keras.datasets import mnist
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense, Flatten
from tensorflow.keras.utils import to_categorical
import time

# 1. Load dataset MNIST (sama seperti sebelumnya)
(X_train, y_train), (X_test, y_test) = mnist.load_data()

# Normalisasi data (0-255 → 0-1)
X_train = X_train / 255.0
X_test = X_test / 255.0

# One-hot encoding label
y_train = to_categorical(y_train, 10)
y_test = to_categorical(y_test, 10)

# 2. Bangun model JST
print("Membangun Model dengan 3 Hidden Layers (512, 256, 128 neuron)...")
model_modified = Sequential([
    Flatten(input_shape=(28, 28)), # Ubah gambar 28x28 menjadi vektor
    Dense(512, activation='relu'), # Hidden layer 1: 512 neuron (Diubah)
    Dense(256, activation='relu'), # Hidden layer 2: 256 neuron (Diubah)
    Dense(128, activation='relu'), # Hidden layer 3: 128 neuron (Ditambahkan)
    Dense(10, activation='softmax') # Output layer (10 kelas)
])

# 3. Kompilasi model
model_modified.compile(optimizer='adam', loss='categorical_crossentropy', metrics=['accuracy'])

# 4. Latih model
print("\nTraining Model Modifikasi...")
start_time = time.time()
# Melatih dengan epochs=10, batch_size=32 (sama seperti model asli)
model_modified.fit(X_train, y_train, epochs=10, batch_size=32, verbose=1)
end_time = time.time()
training_time_modified = end_time - start_time
print(f"Training selesai dalam {training_time_modified:.2f} detik")

# 5. Evaluasi model
print("\nEvaluating Model Modifikasi...")
loss_modified, acc_modified = model_modified.evaluate(X_test, y_test, verbose=0)
print(f"Akurasi Model Modifikasi pada data uji: {acc_modified:.4f}")

Membangun Model dengan 3 Hidden Layers (512, 256, 128 neuron)...

Training Model Modifikasi...


/usr/local/lib/python3.12/dist-packages/keras/src/layers/reshaping/flatten.py:37: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(**kwargs)


Epoch 1/10
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 22s 10ms/step - accuracy: 0.9027 - loss: 0.3261
Epoch 2/10
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 21s 11ms/step - accuracy: 0.9739 - loss: 0.0850
Epoch 3/10
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 20s 11ms/step - accuracy: 0.9803 - loss: 0.0622
Epoch 4/10
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 21s 11ms/step - accuracy: 0.9868 - loss: 0.0430
Epoch 5/10
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 21s 11ms/step - accuracy: 0.9882 - loss: 0.0362
Epoch 6/10
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 20s 11ms/step - accuracy: 0.9918 - loss: 0.0278
Epoch 7/10
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 21s 11ms/step - accuracy: 0.9922 - loss: 0.0233
Epoch 8/10
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 20s 10ms/step - accuracy: 0.9921 - loss: 0.0262
Epoch 9/10
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 23s 12ms/step - accuracy: 0.9933 - loss: 0.0213
Epoch 10/10
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 26s 14ms/step - accuracy: 0.9952 - loss: 0.0164
Training selesai dalam 214.26 detik

Evaluating Model Modifikasi...
Akurasi Model Modifikasi pada d

Model modifikasi dengan tiga hidden layer (512, 256, 128 neuron) mencapai akurasi uji 0.9817 setelah dilatih selama 214.26 detik dan mendapatkan training loss akhir 0.0164.


Eksperimen dengan fungsi aktivasi Sigmoid vs ReLU.

In [4]:
import tensorflow as tf
from tensorflow.keras.datasets import mnist
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense, Flatten
from tensorflow.keras.utils import to_categorical
import time

# 1. Load dataset MNIST (sama seperti sebelumnya)
(X_train, y_train), (X_test, y_test) = mnist.load_data()

# Normalisasi data (0-255 → 0-1)
X_train = X_train / 255.0
X_test = X_test / 255.0

# One-hot encoding label
y_train = to_categorical(y_train, 10)
y_test = to_categorical(y_test, 10)

# 2. Bangun model JST dengan Aktivasi SIGMOID
print("Membangun Model dengan 3 Hidden Layers (512, 256, 128 neuron) menggunakan SIGMOID...")
model_sigmoid = Sequential([
    Flatten(input_shape=(28, 28)), # Ubah gambar 28x28 menjadi vektor
    Dense(512, activation='sigmoid'), # Diubah: ReLU -> Sigmoid
    Dense(256, activation='sigmoid'), # Diubah: ReLU -> Sigmoid
    Dense(128, activation='sigmoid'), # Diubah: ReLU -> Sigmoid
    Dense(10, activation='softmax') # Output layer (tetap softmax)
])

# 3. Kompilasi model
model_sigmoid.compile(optimizer='adam', loss='categorical_crossentropy', metrics=['accuracy'])

# 4. Latih model
print("\nTraining Model dengan SIGMOID...")
start_time = time.time()
# Melatih dengan epochs=10, batch_size=32
model_sigmoid.fit(X_train, y_train, epochs=10, batch_size=32, verbose=1)
end_time = time.time()
training_time_sigmoid = end_time - start_time
print(f"Training selesai dalam {training_time_sigmoid:.2f} detik")

# 5. Evaluasi model
print("\nEvaluating Model dengan SIGMOID...")
loss_sigmoid, acc_sigmoid = model_sigmoid.evaluate(X_test, y_test, verbose=0)
print(f"Akurasi Model Sigmoid pada data uji: {acc_sigmoid:.4f}")

Membangun Model dengan 3 Hidden Layers (512, 256, 128 neuron) menggunakan SIGMOID...

Training Model dengan SIGMOID...


/usr/local/lib/python3.12/dist-packages/keras/src/layers/reshaping/flatten.py:37: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(**kwargs)


Epoch 1/10
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 23s 11ms/step - accuracy: 0.7744 - loss: 0.7284
Epoch 2/10
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 21s 11ms/step - accuracy: 0.9546 - loss: 0.1535
Epoch 3/10
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 21s 11ms/step - accuracy: 0.9700 - loss: 0.0960
Epoch 4/10
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 21s 11ms/step - accuracy: 0.9810 - loss: 0.0630
Epoch 5/10
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 20s 11ms/step - accuracy: 0.9851 - loss: 0.0492
Epoch 6/10
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 21s 11ms/step - accuracy: 0.9899 - loss: 0.0336
Epoch 7/10
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 42s 12ms/step - accuracy: 0.9906 - loss: 0.0284
Epoch 8/10
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 22s 12ms/step - accuracy: 0.9941 - loss: 0.0182
Epoch 9/10
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 21s 11ms/step - accuracy: 0.9949 - loss: 0.0165
Epoch 10/10
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 21s 11ms/step - accuracy: 0.9964 - loss: 0.0122
Training selesai dalam 233.05 detik

Evaluating Model dengan SIGMOID...
Akurasi Model Sigmoid pada 